### PR TITLE
Eng 5025  evaluation set

### PR DESCRIPTION
**Pull request by melisa-writer**
### Adding scripts for test data generation. 
_Remark: I'm not sure where to upload files. It's not a part of the service library -> I've created a subfolder (added to .dockerignore)_

- **get_phrase.py** - looks for verb/noun phrases in pos tags sequences
- **get_options.py** - conjugates verbs (pyinflect lib) or looks for derivationally related forms (NLTK Wordnet)
- **get_gap.py** - uses the above to generate masked sentence, answer and options
- **main.py** - reads csv, applies get_gap to sentences and writes to json
- **utils.py** - helper functions used in main.py (removed from main for clarity)

Sample output: **world_news_in_month_tense.json**

### Updates: 
- dockerignore update (see remark)
- use multi_rank
- GPU support

### CODE REVIEW

File: fitbert_test_data/get_gap.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            answer = chosen_verb_phrase["verb_phrase"].text.lower()
+
+            adverbs = chosen_verb_phrase["adverbs"]
+            verb_phrase_inffix = " ".join([a.text for a in adverbs])
+
+            start = chosen_verb_phrase["start"]
+            end = chosen_verb_phrase["end"]
+
+            options = get_verb_options(main_verb, verb_phrase_inffix)
+            options_len = [1 / len(o) for o in options]
```
Comment (sam-qordoba) [inline fitbert_test_data/get_gap.py]: Can we add some checks around here to make 100% sure we aren't dividing by 0?

File: fitbert/fitb.py [hunk 2]
```diff
@@ -177,12 +160,37 @@ def simplify_options(self, sent: str, options: List[str]):
...
+        ranked = self.rank_multi(sent, options, delemmatize)
+        best_word = ranked[0]
+        return sent.replace(self.mask_token, best_word)
+
+    def mask_fitb(self, sent: str, span: Tuple[int, int]) -> str:
+        masked_str, replaced = self.mask(sent, span)
+        options = [replaced]
+        return self.fitb(masked_str, options, delemmatize=True)
+
+
```
Comment (sam-qordoba) [inline fitbert/fitb.py]: really nit-picky, but the file should end with one blank line.

I use `black` as an auto-formatter and recommend everyone else does, it will automatically fix things like this

File: fitbert_test_data/get_phrase.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+import re
+
+dictionary = {
```
Comment (sam-qordoba) [inline fitbert_test_data/get_phrase.py]: is this format from wordnet?

File: fitbert_test_data/get_options.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+
+
+def get_verb_options(
+    verb, verb_phrase_inffix="", tense=["present", "past", "future"], form=["active"]
+):
+    verb_forms = conjugate(verb)
+    options = []
+    diff_order = ["there", "here", "everywhere", "elsewhere", "anywhere", "nowhere"]
+    for verb_form in verb_forms:
+        if (verb_form[4] in tense) and (verb_form[5] in form):
```
Comment (sam-qordoba) [inline fitbert_test_data/get_options.py]: I really try to not access indices directly - is there a way to make this more literate? e.g. `if is_present_perfect(verb_form):` or whatever it is you are testing?

File: fitbert_test_data/get_options.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+            if verb_phrase_inffix.lower() in diff_order:
+                elements = [verb_phrase_inffix, verb_form[0]] + verb_form[1:3]
```
Comment (sam-qordoba) [inline fitbert_test_data/get_options.py]: Same with this one - can the code use functions or something to make more clear what is happening here?

File: fitbert_test_data/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+- Spacy default model:
+`python3 -m spacy download en_core_web_sm`
+
+- Wordnet for NLTK: 
+```python
+import nltk
+nltk.download()
+```
+
+## Usage
```
Comment (sam-qordoba) [inline fitbert_test_data/README.md]: Is there a command that we could put in this section that will work out of the box?

File: fitbert/fitb.py (hunk 2 above)
Comment (sam-qordoba) [reply fitbert/fitb.py]: This is what my `.vscode/settings.json` looks like (I install certain things like linters and formatters with anaconda, so I don't have to install them in every env):

```json
{
    "python.pythonPath": "venv/bin/python",
    "python.linting.enabled": true,
    "python.linting.mypyEnabled": true,
    "python.linting.mypyPath": "/anaconda3/bin/mypy",
    "python.linting.mypyArgs": [
        "--ignore-missing-imports",
        "--namespace-packages",
        "--follow-imports=silent",
        "--show-column-numbers"
    ],
    "python.linting.pylintEnabled": false,
    "python.formatting.provider": "black",
    "python.formatting.blackPath": "/anaconda3/bin/black",
    "files.exclude": {
        "**/.git": true,
        "**/.svn": true,
        "**/.hg": true,
        "**/CVS": true,
        "**/.DS_Store": true,
        "**/*.pyc": true,
        "**/__pycache__": true,
        "**/.mypy_cache": true
    }
```

**Review by sam-writer — APPROVED** [review]
Approving, because I don't want to block the process, but I do have some things I'd like fixed

File: fitbert_test_data/get_phrase.py (hunk 3 above)
Comment (melisa-qordoba) [reply fitbert_test_data/get_phrase.py]: There are 2 reasons to write custom mapping:

1. Non compatible/obsolete alternatives:

- spacy - only noun chunks

- wordnet - different than spacy pos tag system
https://www.guru99.com/pos-tagging-chunking-nltk.html
https://spacy.io/api/annotation

- textacy - obsolete as we had the following changes:
https://universaldependencies.org/v2/postags.html
CONJ -> CCONJ
In fact also all SCONJ has been mapped to CCONJ (probe on 90k corpus).

2. Possibility to add statistical gaps in the near future: 

Custom mapping allows to find longest common pos tag sequences using Sequence Matcher (https://docs.python.org/2/library/difflib.html).
-> given crowd-sourced examples of fill in the blanks we can find statistically probable places for gaps.
This piece of code has been implemented (and in fact took >75% of coding time), but since no open source examples are available atm - skipped.

File: fitbert_test_data/get_gap.py (hunk 1 above)
Comment (melisa-qordoba) [reply fitbert_test_data/get_gap.py]: There is no chance to get division by zero here. 
See: implementation of conjugate. I am checking whether VB form exists. If VB exist then there are no options of len 0 possible.

File: fitbert/fitb.py (hunk 2 above)
Comment (melisa-qordoba) [reply fitbert/fitb.py]: Good point! I forgot to run black here. Thx!

File: fitbert_test_data/README.md (hunk 6 above)
Comment (melisa-qordoba) [reply fitbert_test_data/README.md]: Nice readme fix, thx!

File: fitbert/fitb.py [hunk 7]
```diff
@@ -12,13 +12,15 @@
 class FitBert:
     def __init__(
         self,
+        device=None,
```
Comment (sam-qordoba) [inline fitbert/fitb.py]: I want these arguments in order of descending-estimated-order-of-usage, but I really can't guess what they will be. Honestly, since there will probably be improvements to BERT in the near future (or [past](https://ai.facebook.com/blog/roberta-an-optimized-method-for-pretraining-self-supervised-nlp-systems/)), maybe we want `model_name` first? I think `device` will be near the end.

File: fitbert/fitb.py [hunk 8]
```diff
@@ -12,13 +12,15 @@
...
         model=None,
         tokenizer=None,
         model_name="bert-large-uncased",
         mask_token="***mask***",
     ):
         self.mask_token = mask_token
         self.delemmatizer = Delemmatizer()
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (sam-qordoba) [inline fitbert/fitb.py]: I think we should check if the user passed in `device` here: if they did, it's probably them trying to tell us to use GPU or CPU. Maybe we should check if they pass `device="cuda"` or `device="cpu"`, and respect that. If neither, check if it is truthy, in which case set `device="cuda"`, and if falsey, set `device="cpu"`?

File: fitbert_test_data/utils.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+def sanitize_text(sentence):
+    for key in abbr_dict:
+        sentence = re.sub(key, abbr_dict[key], sentence, flags=re.IGNORECASE)
+    sentence = sentence[0].upper() + sentence[1:]
+    return sentence
+
+
+def is_valid_sentence(sentence):
+    return not "..." in sentence
```
Comment (sam-qordoba) [inline fitbert_test_data/utils.py]: Is this specific to some of the training data? Because `...` is a valid punctuation mark.

File: fitbert_test_data/get_phrase.py (hunk 3 above)
Comment (sam-qordoba) [reply fitbert_test_data/get_phrase.py]: Can you walk me through this on a call tomorrow?

**Comment by sam-qordoba** [comment]
I added a few more comments, and an update to the main readme. I made a [colab notebook to show usage](https://colab.research.google.com/drive/1NHYRMR7pKt9uY_KfFhIKXwPwvuX-VQLW) and put it in the readme.

I ❤️ this PR. Thank you. I hope you know, the comments all come from a place of loving this code and wanting it to be 💅

File: fitbert_test_data/get_options.py (hunk 4 above)
Comment (melisa-qordoba) [reply fitbert_test_data/get_options.py]: The list has been changed to a dictionary. It looks nicer indeed.

File: fitbert_test_data/get_options.py (hunk 5 above)
Comment (melisa-qordoba) [reply fitbert_test_data/get_options.py]: Changed to filters over dictionaries.

File: fitbert/fitb.py (hunk 8 above)
Comment (melisa-qordoba) [reply fitbert/fitb.py]: Agreed, we can have such fix here. Although I would wait till we finalize 'insides' first, then add options etc.

File: fitbert_test_data/utils.py (hunk 9 above)
Comment (melisa-qordoba) [reply fitbert_test_data/utils.py]: Removing incomplete sentences ending with "...." - condition specific for this corpus, although one could add "ending with "...." condition. Further checking would be required whether its enough (they might end with ... .)
or
you could use a different dataset or preprocess it before running the script.

File: fitbert_test_data/get_phrase.py (hunk 3 above)
Comment (melisa-qordoba) [reply fitbert_test_data/get_phrase.py]: Sure.

File: fitbert/fitb.py (hunk 2 above)
Comment (melisa-qordoba) [reply fitbert/fitb.py]: Done.

File: fitbert_test_data/main.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    dataset_df = dataset_df[0:max_count]
+
+    dataset_with_gaps = []
+
+    for _, row in tqdm(dataset_df.iterrows(), total=len(dataset_df.index)):
+        text = row["text"]
+        doc = nlp(text)
+        for sentence in doc.sents:
+
+            if is_valid_sentence(sentence.text):
```
Comment (sam-qordoba) [inline fitbert_test_data/main.py]: This might be overkill, but just to bring it up so we can talk about it on the call - can we make `is_valid_sentence` somehow a parameter to this script? Since validity checking is based on the dataset?
